core/lib/object_store/src/s3.rs

```diff
@@ -0,0 +1,225 @@
+//! S3-based [`ObjectStore`] implementation.
+
+use std::fmt;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use aws_config::{meta::region::RegionProviderChain, BehaviorVersion, ConfigLoader, Region};
+use aws_runtime::env_config::file::{EnvConfigFileKind, EnvConfigFiles};
+use aws_sdk_s3::{error::SdkError, primitives::ByteStreamError, Client};
+use http::StatusCode;
+
+use crate::raw::{Bucket, ObjectStore, ObjectStoreError};
+
+/// [`ObjectStore`] implementation based on AWS S3.
+pub struct S3Store {
+    endpoint: String,
+    bucket_prefix: String,
+    client: Client,
+}
+
+impl fmt::Debug for S3Store {
+    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
+        formatter
+            .debug_struct("S3Store")
+            .field("bucket_prefix", &self.bucket_prefix)
+            // Skip `client` as its representation may contain sensitive info
+            .finish_non_exhaustive()
+    }
+}
+
+/// Authentication mode for [`S3Store`].
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub enum S3StoreAuthMode {
+    /// Authentication via a credentials file at the specified path.
+    AuthenticatedWithCredentialFile(String),
+    /// Anonymous access (only works for public buckets for read operations).
+    Anonymous,
+}
+
+impl S3Store {
+    /// Creates a new S3 store.
+    pub async fn new(
+        auth_mode: S3StoreAuthMode,
+        bucket_prefix: String,
+        endpoint: Option<String>,
+        region: Option<String>,
+    ) -> Result<Self, ObjectStoreError> {
+        let region_provider = RegionProviderChain::first_try(region.map(Region::new))
+            .or_default_provider()
+            .or_else(Region::new("auto"));
+        let mut sdk_config = Self::get_client_config(auth_mode).region(region_provider);
+        if let Some(endpoint) = endpoint.clone() {
+            tracing::info!(endpoint=%endpoint, "using S3 endpoint defined in storage config");
```

Comment:
Nit: `endpoint=%endpoint` can be shortened to just `%endpoint`.

Reply:
Done